src/Validation/Rules/Leibnitz.elm

```diff
@@ -0,0 +1,277 @@
+module Validation.Rules.Leibnitz exposing (validate, replaceTermWithVar)
+
+import Dict
+import Formula exposing (Formula(..))
+import Formula.Signed exposing (Signed(..))
+import Term exposing (Term(..))
+import Tableau exposing (..)
+import Zipper
+import Validation.Common exposing(..)
+
+
+isEquality : Signed Formula -> Bool
+isEquality f =
+    case f of
+        T (EqAtom lt rt) ->
+            True
+
+        _ ->
+            False
+
+
+leftEqTerm : Signed Formula -> Term
+leftEqTerm f = 
+    case f of 
+        T (EqAtom lt rt) ->
+            lt
+        _ ->
+            Fun "default"[]
+
+
+rightEqTerm : Signed Formula -> Term
+rightEqTerm f = 
+    case f of 
+        T (EqAtom lt rt) ->
+            rt
+        _ ->
+            Fun "default"[]
+
+
+replaceInTerms : Term -> Term -> List Term -> List Term -> Zipper.Zipper -> List Term
+replaceInTerms term var refTerms currentTerms z =
+    case refTerms of
+        [] ->
+            []
+        
+        refTerm :: rts ->
+            case currentTerms of
+                [] ->
+                    refTerms
+
+                currentTerm :: cts ->
+                    (replaceInTerm term var refTerm currentTerm z) ::
+                    (replaceInTerms term var rts cts z) 
+
+
+replaceInTerm : Term -> Term -> Term -> Term -> Zipper.Zipper -> Term
+replaceInTerm term var refTerm currentTerm z =
+    if refTerm == term && refTerm /= currentTerm then  
+        var
+    else
+        case refTerm of
+            Var refStr ->
+                Var refStr
+
+            Fun refStr refTerms ->
+                case currentTerm of
+                    Fun currentStr currentTerms ->
+                        if refStr == currentStr then
+                            Fun refStr (replaceInTerms term var refTerms currentTerms z)
+                        else
+                            Fun refStr refTerms
+                    
+                    Var currentStr ->
+                        var
+
+
+differentStructureError : Zipper.Zipper -> List Problem
+differentStructureError z = 
+    (semanticsProblem z ("The formulas ("
+    ++ (getReffedId Zipper.zSecondRef z) ++ ")"
+    ++ " and (" ++ String.fromInt (z |> Zipper.zNode >> .id) ++ ")" 
+    ++ " should not have different structure"))
+
+
+replaceInFormula : Term -> Term -> Formula -> Formula -> Zipper.Zipper -> 
+ Result (List Problem) (Formula)
+replaceInFormula term var refF currentF z =
```

Comment:
You are trying to do too much in this function. You compare the formulas, and you also check whether the left term in the equality matches the place where the formulas differ.

It would be simpler to have one function `commonTemplate : Formula -> Formula -> Result String Formula` check if the formulas match (or produce `Err` if their structures differ) and produce the common template formula in which `templateVar = Var "[]"` replaces the terms at which the two input formulas differ.

Then two applications of the modified `checkValidSubst` would validate whether the formulas can be obtained by the Leibnitz's rule from one another or not.

I would also advise against dragging a `Zipper` over here. Keep it confined to `check*` functions. You can produce a most important part of the error here and add the references to formulas later.